### PR TITLE
Add Alpha from Grayscale import option to Texture2D

### LINKED ISSUE #5247: Add "Alpha from GrayScale" texture import setting
### Describe the project you are working on

I'm working on a heavily VFX and Particles drived project

### Describe the problem or limitation you are having in your project

I want to use my grayscale texture as the transparency of my particle, but I can't without using a custom shader.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A simple tickbox to choose to use grayscale as alpha or not

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

There should be a boolean in the flags of the texture import settings named "Alpha from Grayscale"

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can be worked around using a custom shader (spatial or canvas_item) like this 

```
shader_type canvas_item;

uniform float _t = 0.5;

void fragment(){
	vec4 tex = texture(TEXTURE, UV);
	if (tex.r < _t && tex.g < _t && tex.b < _t){
		discard;
	}
	COLOR = tex;
}
```

### Is there a reason why this should be core and not an add-on in the asset library?

In my opinion it should be core because every project is using texture assets

### PR DESCRIPTION
**Pull request by Calinou**
This can be used to import textures that are on a solid black background and use them for VFX.

This also adds and exposes `Image::get_format_with_alpha()` and `Image::get_format_without_alpha()` static methods for convenience when working with image formats.

- This closes https://github.com/godotengine/godot-proposals/issues/5247.

**Testing project:** [test_alpha_from_grayscale.zip](https://github.com/godotengine/godot/files/14375387/test_alpha_from_grayscale.zip)
<sub>Sprite from [Kenney Particle Pack](https://kenney.nl/assets/particle-pack), licensed under CC0 1.0 Universal.</sub>

## Preview

alpha_from_grayscale

@Exyde @FyiurAmron Can you confirm this has the expected appearance?

### CODE REVIEW

File: doc/classes/Image.xml [hunk 1]
```diff
@@ -220,6 +220,22 @@
 				Returns the image's format. See [enum Format] constants.
 			</description>
 		</method>
+		<method name="get_format_with_alpha" qualifiers="static">
+			<return type="int" enum="Image.Format" />
+			<param index="0" name="format" type="int" enum="Image.Format" />
+			<description>
+				Returns the equivalent image format to the [param format] parameter specified, but with an alpha channel present. For example, [method get_format_without_alpha] with [constant Image.FORMAT_RGB8] returns [constant Image.FORMAT_RGBA8]. If no equivalent exists, returns the format that was originally passed.
```
Comment (Mickeon) [inline doc/classes/Image.xml:227]: I'd switch the two sentences and put the example on a new line:
```suggestion
				Returns the image format that is equivalent to [param format], but with an alpha channel. If no equivalent exists, returns the format that was originally passed.
				For example, calling this method with [constant Image.FORMAT_RGB8] returns [constant Image.FORMAT_RGBA8].
```

File: doc/classes/Image.xml [hunk 2]
```diff
@@ -220,6 +220,22 @@
...
+				[b]Note:[/b] This method always returns the original format for VRAM-compressed formats.
+			</description>
+		</method>
+		<method name="get_format_without_alpha" qualifiers="static">
+			<return type="int" enum="Image.Format" />
+			<param index="0" name="format" type="int" enum="Image.Format" />
+			<description>
+				Returns the equivalent image format to the [param format] parameter specified, but without an alpha channel present. For example, [method get_format_without_alpha] with [constant Image.FORMAT_RGBA8] returns [constant Image.FORMAT_RGB8]. If no equivalent exists or its equivalent would be larger in memory than the specified format, returns the format that was originally passed.
```
Comment (Mickeon) [inline doc/classes/Image.xml:235]: Same:
```suggestion
				Returns the image format that is equivalent to [param format], but without an alpha channel. If no equivalent exists or the equivalent would occupy more memory than [param format], returns the format that was originally passed.
				For example, calling this method with [constant Image.FORMAT_RGBA8] returns [constant Image.FORMAT_RGB8].
```

File: doc/classes/Image.xml [hunk 3]
```diff
@@ -220,6 +220,22 @@
...
+				[b]Note:[/b] This method always returns the original format for VRAM-compressed formats.
```
Comment (Mickeon) [inline doc/classes/Image.xml:236]: Provide a few examples with "such as _ or _"

**Comment by FyiurAmron** [comment]
@Calinou GIMP's "Color to Alpha" does a different trick, it doesn't apply alpha from grayscale properly. To have it match the reference, you'd have to import the image as its own transparency/alpha mask. Then, the result matches the behaviour of your code perfectly:

image

tl;dr LGTM, your code does exactly what it should.

**Comment by Mickeon** [comment]
> BTW, what's with the brightness on your reference image? It's toned down vs what the actual images from the ZIP have, e.g.

Linear to standard RGB, possibly.

**Comment by Calinou** [comment]
> BTW, what's with the brightness on your reference image? It's toned down vs what the actual images from the ZIP have, e.g.

Filmic tonemapping with whitepoint 6 is used in the MRP.

File: doc/classes/Image.xml (hunk 3 above)
Comment (Calinou) [reply doc/classes/Image.xml]: I've expanded upon the `is_compressed()` documentation while I was at it.

**Review by Mickeon — APPROVED** [review]
That extra detail in `is_compressed` is very, very necessary

**Comment by clayjohn** [comment]
I'm not sure that this is something we need to have in our importer. It seems like a workaround for a very specific workflow. It seems the problem that this is solving is that users have assets that are in grayscale with no alpha since they are designed to be used as additive textures, but they want to use them as an alpha mask without making any changes to the texture in an external program. 

I'm not sure this belongs in Godot asset authoring is something that is better handled by programs dedicated to asset authoring. I think it makes sense to include common operations in Godot or things that are required for places where Godot may differ from the asset authoring tool (like flipping the green channel in normal maps), but I don't think we should add every image operation just because we can. 

The underlying proposal doesn't seem to have gained much interest either in the 2 years its been open. I have a feeling it is something that is very niche.

**Comment by Zireael07** [comment]
> better handled by programs dedicated to asset authoring

The problem with that line of thinking is the same as with thinking "just make assets that line up with Godot's default forward axis". Namely, a lot of devs use freely available assets (e.g. from asset stores, Open Game Art, or other free asset sites) and do NOT have ANY influence on how the asset is made

**Comment by FyiurAmron** [comment]
>  It seems like a workaround for a very specific workflow.

That has been addressed directly in https://github.com/godotengine/godot-proposals/issues/5247#issuecomment-1918198925

> It seems the problem that this is solving is that users have assets that are in grayscale with no alpha since they are designed to be used as additive textures, but they want to use them as an alpha mask without making any changes to the texture in an external program.

See above, ditto.

> I'm not sure this belongs in Godot asset authoring is something that is better handled by programs dedicated to asset authoring.

It, by itself, belongs to neither, and in ideal situation can be handled by either.

>I think it makes sense to include common operations in Godot or things that are required for places where Godot may differ from the asset authoring tool (like flipping the green channel in normal maps), but I don't think we should add every image operation just because we can.

This is not "every image operation". Ditto. Tons of assets and workflows are made with this assumption in mind, and, arguably, that's why e.g. Unity provides this option. Code-wise, it's trivial to implement (as @Calinou clearly showed) - meanwhile, batch-processing the assets to include additional unneeded data (4th channel that is implied from avg. pixel brightness) is tedious and prone to errors.

> The underlying proposal doesn't seem to have gained much interest either in the 2 years its been open. I have a feeling it is something that is very niche.

See above. If 8 people voted on having something that required literally one line of code in a loop and some glue to implement, IMVHO that's the correct ratio of need-vs-effort.

File: doc/classes/Image.xml [hunk 4]
```diff
@@ -79,7 +79,7 @@
 			<param index="1" name="source" type="int" enum="Image.CompressSource" default="0" />
 			<param index="2" name="astc_format" type="int" enum="Image.ASTCFormat" default="0" />
 			<description>
-				Compresses the image to use less memory. Can not directly access pixel data while the image is compressed. Returns error if the chosen compression mode is not available.
+				Compresses the image to use less memory. Cannot directly access pixel data while the image is compressed. Returns error if the chosen compression mode is not available. See also [method is_compressed].
```
Comment (AThousandShips) [inline doc/classes/Image.xml:82]:
```suggestion
				Compresses the image to use less memory. Pixel data cannot directly accessed on a compressed image. Returns error if the chosen compression mode is not available. See also [method is_compressed].
```

**Comment by FyiurAmron** [comment]
@Calinou are there any blockers/prerequisites for this (IIUC, #88763 was, but it's merged already?), or is this ready to be merged?

**Comment by FyiurAmron** [comment]
@Calinou ^ bumping the question above for great justice :)

**Review by BlueCube3310 — APPROVED** [review]
Needs a rebase, but otherwise looks great

**Comment by Calinou** [comment]
Rebased and tested again, it works as expected.

**Review by BlueCube3310 — APPROVED** [review]
Approving for import

**Comment by BlueCube3310** [comment]
Thinking about it after #99676 was merged, this could also be achieved by adding a "Grayscale" option to the remap enum.

**Comment by Repiteo** [comment]
If this approach is desired, it needs a rebase

**Comment by Calinou** [comment]
Superseded by https://github.com/godotengine/godot/pull/99676, as described in https://github.com/godotengine/godot-proposals/issues/5247#issuecomment-4502745889.
